File: fibonacci_algos/fibonacci_huge/fibonacci_huge.py
# Uses python3
import sys
import random

def get_fibonacci_huge_naive(n, m):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m


def get_fibonacci_huge_fast(n, m):
    if n<= 1:
        return n

    fibonacci_mod = [0, 1]

    max_i = 0
    for i in range(2, n+1):
        fibonacci_mod.append((fibonacci_mod[i-1] + fibonacci_mod[i-2])%m)

        max_i = i
        if(fibonacci_mod[i-1] == 0 and fibonacci_mod[i] == 1):
            break

    #print(fibonacci)

    periodiclen = len(fibonacci_mod)
    if max_i < n and max_i != 0:
        periodiclen = periodiclen - 2;

    return (fibonacci_mod[n%periodiclen]%m)


if __name__ == '__main__':

    #Stress test

    #while True:
        #Generate n
        #n = random.randrange(1, 1000000000000000000)

        #Generate m
        #m = random.randrange(2, 100000)

        #print(n, ' ', m)

        #print(get_fibonacci_huge_fast(n, m))


    #input = sys.stdin.read();
    n, m = map(int, input().split())
    print(get_fibonacci_huge_fast(n, m))

File: fibonacci_algos/fibonacci_huge/FibonacciHuge.java
import java.util.*;
import java.math.BigInteger;

public class FibonacciHuge {
    private static long getFibonacciHugeNaive(long n, long m) {
        if (n <= 1)
            return n;

        long previous = 0;
        long current  = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = tmp_previous + current;
        }

        return current % m;
    }

    private static long getFibonacciHugeFast(long n, long m) {
        if ( n <= 1) 
            return n;

        ArrayList<Long> fibonaccimod = new ArrayList<>();

        fibonaccimod.add(0L);
        fibonaccimod.add(1L);
    

        int maxI = 0;

        for (int i = 2; i <= n; i++) {
            long modulo = (fibonaccimod.get(i - 1) + fibonaccimod.get(i - 2))%m;
            fibonaccimod.add(modulo);
        
            //System.out.println("i=" + i + "modulo=" + modulo);
            //periodic.add(modulo);
            //System.out.println(periodic.get(i));
            maxI = i;
            //System.out.println("i=" + i + " period=" + periodic.get(i));
            if(fibonaccimod.get(i-1) == 0 && fibonaccimod.get(i) == 1) {
                break;
            }
        }

        //for (int item : periodic)
            //System.out.print(item + " ");
        //System.out.println("maxI=" + maxI);
        int periodiclen = fibonaccimod.size();
        if ( maxI < n && maxI != 0)
            periodiclen -= 2;
        
        //System.out.println("len=" + periodiclen);
        //int remain = (int) n%periodiclen;
        //System.out.println("remain=" + remain);
        //long remain = (n % (long)periodiclen);
        //System.out.println("remain=" + remain);
        //System.out.println("result=" + fibonaccimod.get((int)(n % periodiclen)));
        return fibonaccimod.get((int)(n%periodiclen));
        //return 5;
    }
    
    public static void main(String[] args) {
        //Testing

       /* while(true) {
            Random rand = new Random();
            //Generate n
            long n = Math.abs(rand.nextLong()) % 1000000000000000000 + 1;

            //Generate m
            long m = Math.abs(rand.nextLong()) % 100000 + 2;
            System.out.println(n + " " + m);

            System.out.println(getFibonacciHugeFast(n, m));
        }*/

        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long m = scanner.nextLong();
        System.out.println(getFibonacciHugeFast(n, m));
    }
}

